testdata/p4_16_samples_outputs/p416-type-use3.p4.p4info.txt

```diff
@@ -0,0 +1,261 @@
+pkg_info {
+  arch: "v1model"
+}
+tables {
+  preamble {
+    id: 33586815
+    name: "ingress.custom_table"
+    alias: "custom_table"
+  }
+  match_fields {
+    id: 1
+    name: "hdr.ethernet.srcAddr"
+    bitwidth: 48
+    match_type: EXACT
+    type_name {
+      name: "EthT_t"
+    }
+  }
+  match_fields {
+    id: 2
+    name: "hdr.ethernet.dstAddr"
+    bitwidth: 48
+    match_type: EXACT
+  }
+  match_fields {
+    id: 3
+    name: "hdr.custom.e"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 4
+    name: "hdr.custom.ed"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 5
+    name: "hdr.custom.et"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomT_t"
+    }
+  }
+  match_fields {
+    id: 6
+    name: "hdr.custom.edd"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 7
+    name: "hdr.custom.eddt"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomDDT_t"
+    }
+  }
+  match_fields {
+    id: 8
+    name: "hdr.custom.edtd"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomDT_t"
+    }
+  }
+  match_fields {
+    id: 9
+    name: "hdr.custom.edtt"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomDTT_t"
+    }
+  }
+  match_fields {
+    id: 10
+    name: "hdr.custom.etdd"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomT_t"
+    }
+  }
+  match_fields {
+    id: 11
+    name: "hdr.custom.etdt"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomT_t"
+    }
+  }
+  match_fields {
+    id: 12
+    name: "hdr.custom.ettd"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomT_t"
```

Comment:
And this one should be "CustomTT_t".  Several others below look incorrect to me.  This PR contains the latest version of the guidelines for generating P4Info.  They have not changed in any substantial way since the last version I sent out a link for, except there are a few open questions about how to handle serializable enum types that I do not know the answer for yet.

Reply:
This is the PR I meant to link to in the previous comment: https://github.com/p4lang/p4runtime/pull/193/files

Reply:
Thanks, Andy.  I checked in p4c code changes that passed results for what you asked of.

@antoninbas New getTypeName() has been checked in.  

Again, I have punted type_info generation for table key and Action param to a subsequent PR.  If all is well, I will close this PR later in the evening.  Then I can issue a new PR.  thanks all!

Reply:
type_name values look better to me now.

Some of the bitwidth values are, I think, incorrect in some cases when there are p4runtime_translation annotations present on the types, but perhaps those fixes can wait for a future PR, depending upon what others think.

Reply:
Thanks for reviewing, Andy.  I am taking a look at any incorrect bitwidth in all eight files I use to test this PR and will try to fix them tonight.

Reply:
I have not yet reviewed your p4info output files to see if there are incorrect bitwidths there.

In general, if a value has at least one `type` definition involved in its type, take the one 'closest' to the value (the first type in the type_list(variable) list, in the terms of my writeup on generating P4info files), and if it has a `p4runtime_translation` annotation, then the `bitwidth` value must come from the second argument of the annotation, not from the underlying type.

Reply:
Right - this is my understanding.  Before any checkin, I use a python script that tests 8 files which are impacted directly by this PR.  The files are:
```
p4/testdata/p4_16_samples/p416-type-use3.p4
p4/testdata/p4_16_samples/psa-example-counters-bmv2.p4
p4/testdata/p4_16_samples/psa-example-digest-bmv2.p4
p4/testdata/p4_16_samples/psa-example-register2-bmv2.p4
p4/testdata/p4_16_samples/psa-portid-using-newtype2.p4
p4/testdata/p4_16_samples/v1model-digest-containing-ser-enum.p4
p4/testdata/p4_16_samples/table-key-serenum.p4
p4/testdata/p4_16_samples/p4rt_digest_complex.p4
```

I checked bitwidth in the p4info.txt of these files and here is the result.

```
p416-type-use3.p4 - todo
psa-example-counters-bmv2.p4 - good
psa-example-digest-bmv2.p4 - good
psa-example-register2-bmv2.p4 - good
psa-portid-using-newtype2.p4 - does not have p4info.txt file.
v1model-digest-containing-ser-enum.p4 - good
table-key-serenum.p4   - good
p4rt_digest_complex.p4 - good
```

Reply:
I completed checking p4info.txt for p416-type-use3.p4 and did not find any incorrect bitwidth.  No rush, when you get time, please let me know any other p4info.txt which has incorrect bitwidth or if I missed anything in my checks - thanks.

Reply:
If you want another torture test that I think your latest code may have trouble with some bitwidths because of annotations, here is the P4 program: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/v1model-p4runtime-most-types1.p4

and here is what I think is at least close to a correct P4Info output file generated from it should be: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/expected-p4info-files/v1model-p4runtime-most-types1.v1model.p4info.txt

Reply:
Sure, thanks.  The more testing, the better.  I have generated p4info.txt and will inspect it.

Reply:
@antoninbas The p4c p4runtime typeSpecConverter only works with `type` if the `type` exists in psa.p4.  User-defined `type`s, in a P4 program like what Andy just gave me, don't invoke the 'TypeSpecConverter::preorder(const IR::Type_Newtype* type)` method.

Reply:
My question was to understand the software better.  In the meantime, I added checking for annotation in p4RuntimeSerializer and now the user-defined annotations are being processed.

Reply:
> If you want another torture test that I think your latest code may have trouble with some bitwidths because of annotations, here is the P4 program: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/v1model-p4runtime-most-types1.p4
> 
> and here is what I think is at least close to a correct P4Info output file generated from it should be: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/expected-p4info-files/v1model-p4runtime-most-types1.v1model.p4info.txt

If you see your p4info.txt, where you have a "48" it should be 49 because the annotation says 49.  Likewise, for Custom2_t, bitwidth is 12 in annotation, but your p4info.txt uses 8.

Reply:
I agree with your corrections, after comparing my expected p4info file to the one you now have in this PR.  I have updated mine to match the corrections you have pointed out.  There are still missing things for the controller_header in my P4 code, but that can wait for a future PR, as far as I am concerned.  Thanks for the corrections.

Reply:
The controller_header annotation is legacy code which I haven't touched.  When you complete controller_header in your P4 code, one can look at any issues that arise.  Thanks for all the great P4 programs to shake out p4c feature development.

Tomorrow, I plan to create a new PR with consolidated diffs so that review is easier for reviewers.